### PR TITLE
fix: nullptr check when closing windows

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/14161.

We weren't performing a nullptr check, so we would experience an `EXC_BAD_ACCESS` crash
if an object in the WindowVector is destroyed during cleanup. This fixes that by adding said nullptr check.

cc @zcbenz @deepak1556 @MarshallOfSound 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed an occasional crash when closing all BrowserWindows.

### CODE REVIEW

**Comment by deepak1556** [comment]
changing to `std::vector<base::WeakPtr<NativeWindow>>` will help rather than maintaining a raw map, the ideal solution is for the map to be maintained by `AddWindow` , `RemoveWindow` , but I don't know when is case this happens. Open to solutions.

**Comment by MarshallOfSound** [comment]
If we make this a vector of weak pointers this solution makes sense, without it I don't think this fix actually helps.

**Comment by deepak1556** [comment]
@zcbenz any reason not maintain the `WindowList` vector  as weakptr rather than converting at the call site ?

**Comment by zcbenz** [comment]
I think the internal records should still be a vector of raw ptrs, the purpose of `WindowList` is to provide a precise list of existing windows and it is doing the work correctly. Changing the internal vector to store weak ptrs would be a weird design, because `WindowList` already correctly manages the window list so there is no way for its elements to be null. 

There is crash because `CloseAllWindows` and `DestroyAllWindows` are deleting windows while iterating them, so I think we should only fix these methods.

File: shell/browser/window_list.cc [hunk 1]
```diff
@@ -84,16 +96,22 @@ void WindowList::CloseAllWindows() {
 #if defined(OS_MACOSX)
   std::reverse(windows.begin(), windows.end());
 #endif
-  for (auto* const& window : windows)
-    if (!window->IsClosed())
+  for (auto* const& window : windows) {
```
Comment (deepak1556) [inline shell/browser/window_list.cc:99]: Weakptr vector here too

File: shell/browser/window_list.cc [hunk 2]
```diff
@@ -81,19 +93,27 @@ void WindowList::RemoveObserver(WindowListObserver* observer) {
 // static
 void WindowList::CloseAllWindows() {
   WindowVector windows = GetInstance()->windows_;
+  std::vector<base::WeakPtr<NativeWindow>> weak_windows =
+      ConvertToWeakPtrVector(windows);
```
Comment (zcbenz) [inline shell/browser/window_list.cc:97]: `WindowVector windows = GetInstance()->windows_` is doing a copy of vector, can just pass `ConvertToWeakPtrVector(GetInstance()->windows_)` here to save the copy.

File: shell/browser/window_list.cc [hunk 3]
```diff
@@ -10,6 +10,18 @@
 #include "shell/browser/native_window.h"
 #include "shell/browser/window_list_observer.h"
 
+namespace {
+template <typename T>
+std::vector<base::WeakPtr<T>> ConvertToWeakPtrVector(std::vector<T*> raw_ptrs) {
+  std::vector<base::WeakPtr<T>> converted_to_weak;
```
Comment (zcbenz) [inline shell/browser/window_list.cc:16]: Can do a `converted_to_weak.reserve(raw_ptrs.size())` here;

File: shell/browser/window_list.cc [hunk 4]
```diff
@@ -10,6 +10,18 @@
...
+  for (auto* raw_ptr : raw_ptrs) {
+    base::WeakPtr<T> weak = raw_ptr->GetWeakPtr();
+    converted_to_weak.push_back(weak);
```
Comment (zcbenz) [inline shell/browser/window_list.cc:19]: Can do `push_back(raw_ptr->GetWeakPtr()` so it is moved instead of copied.

**Pull request merged**